### PR TITLE
GAWB-23: Create endpoints for importing and exporting workspace attributes as tsv files

### PR DESCRIPTION
**Pull request by ansingh7115**
- [x] **Submitter**: Include the JIRA issue number in the PR description
- [x] **Submitter**: Make sure Swagger is updated if API changes
- [x] **Submitter**: If updating admin endpoints, also update [firecloud-admin-cli](https://github.com/broadinstitute/firecloud-admin-cli)
- [ ] **Submitter**: Check documentation and code comments. Add explanatory PR comments if helpful.
- [x] **Submitter**: JIRA ticket checks:
  * Acceptance criteria exists and is met
  * Note any changes to implementation from the description
  * Add notes on what you've tested
- [x] **Submitter**: Update RC_XXX release ticket with any config or environment changes necessary
- [x] **Submitter**: Update FISMA documentation if changes to:
  * Authentication
  * Authorization
  * Encryption
  * Audit trails
- [x] **Submitter**: If you're adding new libraries, sign us up to security updates for them
- [ ] Tell the tech lead (TL) that the PR exists if they wants to look at it
- [x] Anoint a lead reviewer (LR). **Assign PR to LR**
* Review cycle:
  * LR reviews
  * Rest of team may comment on PR at will
  * **LR assigns to submitter** for feedback fixes
  * Submitter rebases to develop again if necessary
  * Submitter makes further commits. DO NOT SQUASH
  * Submitter updates documentation as needed
  * Submitter **reassigns to LR** for further feedback
- [ ] **TL** sign off
- [ ] **LR** sign off
- [ ] **Assign to submitter** to finalize
- [ ] **Submitter**: Verify all tests go green, including CI tests
- [ ] **Submitter**: Squash commits and merge to develop
- [ ] **Submitter**: Delete branch after merge
- [ ] **Submitter**: **Test this change works on dev environment after deployment**. YOU own getting it fixed if dev isn't working for ANY reason!
- [ ] **Submitter**: Verify swagger UI on dev environment still works after deployment
- [ ] **Submitter**: Inform other teams of any API changes via hipchat and/or email
- [ ] **Submitter**: Mark JIRA issue as resolved once this checklist is completed

### CODE REVIEW

File: src/main/scala/org/broadinstitute/dsde/firecloud/service/WorkspaceService.scala [hunk 1]
```diff
@@ -61,4 +71,71 @@ class WorkspaceService(protected val argUserInfo: UserInfo, val rawlsDAO: RawlsD
...
+          case Success(attrs) =>
+              rawlsDAO.getWorkspace(workspaceNamespace, workspaceName) flatMap { workspaceResponse =>
+                  val allOperations = getWorkspaceAttributeCalls(attributePairs)
+                  rawlsDAO.patchWorkspaceAttributes(workspaceNamespace, workspaceName, allOperations) map (RequestComplete(_))
+                }
+        }
+      }
+  }
+
+  private def getWorkspaceAttributeCalls(attributePairs: Seq[(String,String)]): Seq[AttributeUpdateOperation] = {
```
Comment (MatthewBemis) [inline src/main/scala/org/broadinstitute/dsde/firecloud/service/WorkspaceService.scala:126]: for readability, you can name the tuple fields using `case` like so:

```
    attributePairs.map { case (name, value) =>
      if (value.equals("__DELETE__")) {
        new RemoveAttribute(new AttributeName("default", name))
      }
      else {
        new AddUpdateAttribute(new AttributeName("default", name), new AttributeString(value))
      }
    }
```

the same thing applies to exportWorkspaceAttributes above

File: src/test/scala/org/broadinstitute/dsde/firecloud/service/WorkspaceApiServiceSpec.scala [hunk 2]
```diff
@@ -549,6 +578,17 @@ class WorkspaceApiServiceSpec extends BaseServiceSpec with WorkspaceApiService w
           }
         }
       }
+
+      "when calling POST on the workspaces/*/*/importAttributes path" - {
+        "should 200 OK if it has the correct headers and valid internals" in {
```
Comment (MatthewBemis) [inline src/test/scala/org/broadinstitute/dsde/firecloud/service/WorkspaceApiServiceSpec.scala:583]: maybe also test a failure case or two here? failing the check in `checkFirstRowDistinct` may be a good one

File: src/main/scala/org/broadinstitute/dsde/firecloud/service/WorkspaceService.scala [hunk 3]
```diff
@@ -61,4 +71,71 @@ class WorkspaceService(protected val argUserInfo: UserInfo, val rawlsDAO: RawlsD
...
+  }
+
+   /*
+  Import attributes on a workspace from a TSV file
+  */
+  private def importWorkspaceAttributeTSV(workspaceNamespace: String, workspaceName: String, tsv: TSVLoadFile): Future[PerRequestMessage] = {
+    checkFirstRowDistinct(tsv) {
+        val attributePairs = colNamesToWorkspaceAttributeNames(tsv.headers).zip(tsv.tsvData.head)
+        Try(scala.util.parsing.json.JSONObject(attributePairs.toMap).toString().parseJson.asJsObject.convertTo[AttributeMap]) match {
+          case Failure(ex:ParsingException) => Future(RequestCompleteWithErrorReport(StatusCodes.BadRequest, "Invalid json supplied", ex))
```
Comment (MatthewBemis) [inline src/main/scala/org/broadinstitute/dsde/firecloud/service/WorkspaceService.scala]: if this is an error that the user is going to see, "Invalid json supplied" will not make a whole lot of sense to them because they're uploading a tsv

File: src/main/scala/org/broadinstitute/dsde/firecloud/service/WorkspaceService.scala [hunk 4]
```diff
@@ -61,4 +71,85 @@ class WorkspaceService(protected val argUserInfo: UserInfo, val rawlsDAO: RawlsD
     }
   }
 
+  def exportWorkspaceAttributes(workspaceNamespace: String, workspaceName: String, filename: String): Future[PerRequestMessage] = {
+    log.info("WE'RE DOING STUFF!!!")
```
Comment (MatthewBemis) [inline src/main/scala/org/broadinstitute/dsde/firecloud/service/WorkspaceService.scala]: ❗️❗️

File: src/main/scala/org/broadinstitute/dsde/firecloud/service/WorkspaceService.scala [hunk 5]
```diff
@@ -61,4 +71,85 @@ class WorkspaceService(protected val argUserInfo: UserInfo, val rawlsDAO: RawlsD
...
+    Try(rawlsDAO.getWorkspace(workspaceNamespace, workspaceName)) match {
+      case Failure(regret) => Future(RequestCompleteWithErrorReport(StatusCodes.BadRequest, regret.getMessage))
+      case Success(workspaceFuture) => workspaceFuture map { workspaceResponse =>
+          val headerString = "workspace:" + (workspaceResponse.workspace.get.attributes map { attribute =>
```
Comment (MatthewBemis) [inline src/main/scala/org/broadinstitute/dsde/firecloud/service/WorkspaceService.scala]: you can also do the tuple field naming trick here

File: src/main/scala/org/broadinstitute/dsde/firecloud/service/WorkspaceService.scala [hunk 6]
```diff
@@ -61,4 +71,85 @@ class WorkspaceService(protected val argUserInfo: UserInfo, val rawlsDAO: RawlsD
...
+    }
+  }
+
+  private def colNamesToWorkspaceAttributeNames(headers: Seq[String]): Seq[String] = {
+    val newHead = headers.head.stripPrefix("workspace:")
+    Seq(newHead) ++ headers.tail
+  }
+
+  private def checkIf2Rows(tsv: TSVLoadFile)(op: => Future[PerRequestMessage]): Future[PerRequestMessage] = {
+    log.info(tsv.tsvData.toString())
```
Comment (MatthewBemis) [inline src/main/scala/org/broadinstitute/dsde/firecloud/service/WorkspaceService.scala]: is this log statement meant to be left in?

File: src/main/scala/org/broadinstitute/dsde/firecloud/service/WorkspaceService.scala (hunk 4 above)
Comment (ansingh7115) [reply src/main/scala/org/broadinstitute/dsde/firecloud/service/WorkspaceService.scala]: Crap, I totally forgot to take that out

File: src/main/scala/org/broadinstitute/dsde/firecloud/service/WorkspaceService.scala [hunk 7]
```diff
@@ -61,4 +71,83 @@ class WorkspaceService(protected val argUserInfo: UserInfo, val rawlsDAO: RawlsD
     }
   }
 
+  def exportWorkspaceAttributes(workspaceNamespace: String, workspaceName: String, filename: String): Future[PerRequestMessage] = {
+    Try(rawlsDAO.getWorkspace(workspaceNamespace, workspaceName)) match {
+      case Failure(regret) => Future(RequestCompleteWithErrorReport(StatusCodes.BadRequest, regret.getMessage))
```
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/firecloud/service/WorkspaceService.scala:78]: 2 comments: 1) you don't need a full fledged Future for something we already know the value of. In this case Future.successful creates something that looks like a Future but completes immediately. 2) Why is this a bad request? Seems like you are masking the underlying error. Maybe it was a 404? Maybe it was a communications error?

Comment (dvoet) [reply src/main/scala/org/broadinstitute/dsde/firecloud/service/WorkspaceService.scala:78]: why do you even need to handle the error here at all? Is there something special about the error handling that is different than what per request would do?

Comment (dvoet) [reply src/main/scala/org/broadinstitute/dsde/firecloud/service/WorkspaceService.scala:78]: and 1 more: this is probably not doing what you think it is. The Try is not wrapping the actual call to rawls. It is wrapping only the construction of the call. rawlsDAO.getWorkspace(workspaceNamespace, workspaceName) creates a Future, the Try wraps the creation of the Future, not the actual execution. So calling this error a bad request is wrong, it is an internal server error.

TLDR; get rid of the Try

File: src/main/scala/org/broadinstitute/dsde/firecloud/service/WorkspaceService.scala [hunk 8]
```diff
@@ -61,4 +71,83 @@ class WorkspaceService(protected val argUserInfo: UserInfo, val rawlsDAO: RawlsD
...
+      case Success(workspaceFuture) => workspaceFuture map { workspaceResponse =>
+          val headerString = "workspace:" + (workspaceResponse.workspace.get.attributes map { case (attName, attValue) =>
+                attName.name}).mkString("\t").replaceAll("description\t", "")
```
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/firecloud/service/WorkspaceService.scala]: how are namespaces handled? Looks like they are completely dropped. Will that be a problem for library things?

File: src/main/scala/org/broadinstitute/dsde/firecloud/service/WorkspaceService.scala [hunk 9]
```diff
@@ -61,4 +71,83 @@ class WorkspaceService(protected val argUserInfo: UserInfo, val rawlsDAO: RawlsD
...
+          val valueString = (workspaceResponse.workspace.get.attributes map { attribute =>
+              impAttributeFormat.write(attribute._2).toString().replaceAll("\"","")}).mkString("\t").replaceAll("null\t", "")
```
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/firecloud/service/WorkspaceService.scala]: what is attribute._2? replace attribute 1 line up with a case statement that extracts the tuple to make the code more clear

replacing null\t with "", is it right to drop the tab?

File: src/main/scala/org/broadinstitute/dsde/firecloud/service/WorkspaceService.scala [hunk 10]
```diff
@@ -61,4 +71,83 @@ class WorkspaceService(protected val argUserInfo: UserInfo, val rawlsDAO: RawlsD
...
+          RequestComplete(StatusCodes.OK, headerString + "\n" + valueString)
+        }
+    }
+  }
+
+  def importAttributesFromTSV(workspaceNamespace: String, workspaceName: String, tsvString: String): Future[PerRequestMessage] = {
+    Try(TSVParser.parse(tsvString)) match {
+      case Failure(regret) => Future(RequestCompleteWithErrorReport(StatusCodes.BadRequest, regret.getMessage))
```
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/firecloud/service/WorkspaceService.scala:95]: Future.successful

File: src/main/scala/org/broadinstitute/dsde/firecloud/service/WorkspaceService.scala [hunk 11]
```diff
@@ -61,4 +71,83 @@ class WorkspaceService(protected val argUserInfo: UserInfo, val rawlsDAO: RawlsD
...
+
+   /*
+  Import attributes on a workspace from a TSV file
+  */
+  private def importWorkspaceAttributeTSV(workspaceNamespace: String, workspaceName: String, tsv: TSVLoadFile): Future[PerRequestMessage] = {
+    checkIf2Rows(tsv) {
+      checkFirstRowDistinct(tsv) {
+        val attributePairs = colNamesToWorkspaceAttributeNames(tsv.headers).zip(tsv.tsvData.head)
+        Try(scala.util.parsing.json.JSONObject(attributePairs.toMap).toString().parseJson.asJsObject.convertTo[AttributeMap]) match {
+          case Failure(ex: ParsingException) => Future(RequestCompleteWithErrorReport(StatusCodes.BadRequest, "Invalid json supplied", ex))
```
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/firecloud/service/WorkspaceService.scala:114]: how could this happen? The user is not even giving us json so telling them invalid json does not seem helpful.

File: src/main/scala/org/broadinstitute/dsde/firecloud/webservice/WorkspaceApiService.scala [hunk 12]
```diff
@@ -85,6 +101,26 @@ trait WorkspaceApiService extends HttpService with FireCloudRequestBuilding
                     }
                   }
                 } ~
+                path("exportAttributes") {
```
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/firecloud/webservice/WorkspaceApiService.scala:103]: why are there 2 exportAttributes end points?

**Comment by dvoet** [comment]
need export tests
